File: 0002_codes/C - Kernel_Level/OS03.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#if defined(_WIN32) || defined(_WIN64)
    #include <direct.h>
    #define get_cwd _getcwd
    #define change_dir _chdir
#else
    #include <unistd.h>
    #define get_cwd getcwd
    #define change_dir chdir
#endif

/*
 * OS03 : working directory navigator
 *
 * This program:
 * - prints the current working directory
 * - asks the user for a new directory
 * - tries to change to the new directory
 * - prints the updated working directory or an error message
 */

int main(void) {
    char current_dir[1024];
    char new_dir[1024];

    printf("=========================================\n");
    printf("           OS03 - working dir            \n");
    printf("=========================================\n");

    /* get and print current working directory */
    if (get_cwd(current_dir, sizeof(current_dir)) == NULL) {
        printf("could not get current directory\n");
        printf("=========================================\n");
        return 1;
    }

    printf("current directory:\n%s\n\n", current_dir);

    /* ask user for new directory */
    printf("enter new directory (or '.' to stay here):\n");
    if (fgets(new_dir, sizeof(new_dir), stdin) == NULL) {
        printf("input error\n");
        printf("=========================================\n");
        return 1;
    }

    /* remove newline at the end if present */
    char *newline = strchr(new_dir, '\n');
    if (newline != NULL) {
        *newline = '\0';
    }

    /* if user entered '.' or empty string, do nothing */
    if (new_dir[0] == '\0' || (new_dir[0] == '.' && new_dir[1] == '\0')) {
        printf("\nno change, staying in the same directory\n");
        printf("=========================================\n");
        return 0;
    }

    /* try to change directory */
    if (change_dir(new_dir) != 0) {
        printf("\ncould not change directory to:\n%s\n", new_dir);
        printf("check if the path exists and you have access\n");
        printf("=========================================\n");
        return 1;
    }

    /* print updated directory */
    if (get_cwd(current_dir, sizeof(current_dir)) == NULL) {
        printf("\nchanged directory, but could not read new path\n");
        printf("=========================================\n");
        return 1;
    }

    printf("\nnew current directory:\n%s\n", current_dir);
    printf("=========================================\n");

    return 0;
}

File: 0002_codes/JAVA - User_Level/OS03.java
import java.io.File;
import java.util.Scanner;

/*
 * OS03 - working directory navigator
 *
 * this program:
 * - prints the current working directory
 * - asks the user for a new directory
 * - tries to change to the new directory (inside the program)
 * - prints the updated directory or an error message
 *
 * NOTE: java cannot change the real working directory of the system shell,
 * so we simulate the change by tracking our own "current" directory.
 * Thats why I preffer you all learn C (say C is best when u meet me so i will know u saw this)
 */

public class OS03 {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        /* get current working directory from system property */
        String currentDir = System.getProperty("user.dir");

        System.out.println("=========================================");
        System.out.println("           OS03 - working dir            ");
        System.out.println("=========================================");

        System.out.println("current directory:");
        System.out.println(currentDir + "\n");

        System.out.print("enter new directory (or '.' to stay here): ");
        String newDir = scanner.nextLine().trim();

        if (newDir.isEmpty() || newDir.equals(".")) {
            System.out.println("\nno change, staying in the same directory");
            System.out.println("=========================================");
            return;
        }

        /* create file object for the target directory */
        File target = new File(newDir);

        /* if the path is not absolute, resolve it relative to currentDir */
        if (!target.isAbsolute()) {
            target = new File(currentDir, newDir);
        }

        /* check if directory exists and is accessible */
        if (!target.exists() || !target.isDirectory()) {
            System.out.println("\ncould not change directory to:");
            System.out.println(target.getAbsolutePath());
            System.out.println("check if the path exists and you have access");
            System.out.println("=========================================");
            return;
        }

        /* simulate changing directory */
        currentDir = target.getAbsolutePath();

        System.out.println("\nnew current directory:");
        System.out.println(currentDir);
        System.out.println("=========================================");
    }
}
